jcl/src/java.base/share/classes/openj9/internal/foreign/abi/InternalDowncallHandler.java

```diff
@@ -89,6 +90,19 @@ public class InternalDowncallHandler {
 	private MemoryLayout realReturnLayout;
 	private MethodHandle boundMH;
 
+	/*[IF JAVA_SPEC_VERSION >= 22]*/
+	/* The identifier denotes the passed-in argument is a heap segment which helps
+	 * extract the address from the heap argument's base object in native.
+	 */
+	private static final long DOWNCALL_HEAP_ARGUMENT_ID = 0x1;
+
+	/* The array stores the base object (which holds the heap address as defined in OpenJDK)
+	 * of the heap arguments so as to extract the heap address from the base object in native.
+	 */
+	private Object[] heapBases;
+	private int curHeapArgIdx;
```

Comment:
This doesn't seem to be thread-safe. Multiple threads might concurrently invoke the same method handle referring to an instance of this class: they will clash over updates to these fields, including the potential for `ArrayIndexOutOfBoundsException`.

Reply:
Agreed & I will update the related code to deal with the case of multiple threads.

Reply:
I updated the code by introducing `ConcurrentHashMap` to hold the heap base array plus the corresponding offset array for the current thread to avoid the issue with multi-threading (verified in internal  & external tests).